crate/operator/utils/kopf.py

```diff
@@ -42,45 +43,75 @@ async def _wrapper(**_):
     return _wrapper
 
 
+class Notification(TypedDict):
+    event: WebhookEvent
+    payload: WebhookSubPayload
+    status: WebhookStatus
+
+
 class StateBasedSubHandler(abc.ABC):
     """
-    State-based sub-handlers allow running sub-handlers in a defined order.
-
-    Kopf sub-handlers run in parallel, meaning, the order of them being
-    executed is not strictly defined. When updating a custom CrateDB resource,
-    though, the order of the steps to deal with the changes in the operator are
-    predefined.
-
-    Subclasses of :class:`StateBasedSubHandler` are required to implement the
-    :meth:`~StateBasedSubHandler.handle` method, which is the code that'll run
-    as a sub-handler. The class will ensure that the sub-handler is only run
-    when the current state in :class:`the statemachine
-    <crate.operator.utils.state.StateMachine>` is :attr:`state`.
-    """
+    A handler capable of waiting for other handlers to finish.
 
-    #: The :class:`~crate.operator.utils.state.State` at which to run the
-    #: subhandler. Until the state has been reached, the sub-handler will be
-    #: retried using :class:`kopf.TemporaryError`.
-    state: State
+    This can be expressed as a set of dependencies passed to the ``depends_on``
+    parameter of the constructor.
+
+    Basically, we wrap an actual handler here. Before executing the actual one,
+    we check if the dependencies have completed yet - we do this by checking
+    the statuses of those dependent handlers, as stored in the ``status`` field
+    of the CrateDB resource.
+    """
 
-    def __init__(self, namespace: str, name: str, context: Context):
+    def __init__(
+        self,
+        namespace: str,
+        name: str,
+        ref: str,
+        context: dict,
+        depends_on=None,
+        run_on_dep_failures=False,
+    ):
+        """
+        Constructs a new dependency-aware handler.
+
+        :param namespace: the namespace to use
+        :param name: the name of the CrateDB resource we're working on
+        :param context: _deprecated_
+        :param depends_on: list of dependent handler this handler should wait for.
+        :param run_on_dep_failures: whether we should still execute if our dependencies
+            have failed. This is useful for handlers that clean up resources after
+            other handlers have finished, and always need to run.
+        """
         self.namespace = namespace
         self.name = name
+        self.ref = ref
         self._context = context
+        self.depends_on = depends_on if depends_on is not None else []
```

Comment:
Good point, but I did it this way since PyCharm complained:
![Screenshot from 2021-04-08 16-09-08](https://user-images.githubusercontent.com/4594061/114051182-c7324e80-9884-11eb-8d2e-6e50f752c4dc.png)

:shrug: ?

Reply:
Okay, this might have been a bad suggestions, let's not use mutable values as default values, especially not lists as this might cause issues later on 😅